### PR TITLE
cpu: saml21 initial commit

### PR DESCRIPTION
**Pull request by kaspar030**
basic port, uart, one timer, gpio, spi working.

Needs MCU support patch for openocd. [1] 

My work on this port was mostly paid by FreshTemp, LLC. Thanks, guys!

[1] https://www.mail-archive.com/openocd-devel@lists.sourceforge.net/msg07241.html

### CODE REVIEW

**Comment by OlegHahm** [comment]
> +21,943 −0 

Spammer! :stuck_out_tongue_closed_eyes:

**Comment by kaspar030** [comment]
have to top your codestyle commits somehow! ;)

**Comment by OlegHahm** [comment]
By importing hundreds of vendor headers? *j/k*

**Comment by kaspar030** [comment]
Is that an ACK? ;)

**Comment by OlegHahm** [comment]
GitHub does not even allow me to review:
> Sorry, we could not display the entire diff because it was too big. 

(Best excuse ever...)

**Comment by kaspar030** [comment]
If I'm the only one with that board, who's gonna review this?

**Comment by OlegHahm** [comment]
Maybe at the Hack'n'ACK someone can review?

File: boards/saml21-xpro/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# define the cpu used by the arduino due board
```
Comment (haukepetersen) [inline boards/saml21-xpro/Makefile.include]: s/arduino due/saml21-xpro/

File: boards/saml21-xpro/board.c [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    /* initialize the boards LEDs */
+    led_init();
+}
+
+
+/**
+ * @brief Initialize the boards on-board LED
+ */
+void led_init(void)
```
Comment (haukepetersen) [inline boards/saml21-xpro/board.c]: you could just throw out this function...

File: boards/saml21-xpro/include/board.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ */
+#define LED_PORT            PORT->Group[0]
+#define LED_PIN             (19)
+/** @} */
+
+/**
+ * @name Macros for controlling the on-board LEDs.
+ * @{
+ */
+#define LED_ON              (LED_PORT.OUTCLR.reg = 1<<LED_PIN)
```
Comment (haukepetersen) [inline boards/saml21-xpro/include/board.h]: in the board.h you did not initialize any LED -> does the board have LEDs?

File: boards/saml21-xpro/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+ */
+#define TIMER_NUMOF        (1U)
+#define TIMER_0_EN         1
+
+/* Timer 0 configuration */
+#define TIMER_0_DEV        TC0->COUNT32
+#define TIMER_0_CHANNELS   1
+#define TIMER_0_MAX_VALUE  (0xffffffff)
+#define TIMER_0_ISR        isr_tc0
+
```
Comment (haukepetersen) [inline boards/saml21-xpro/include/periph_conf.h]: newline

File: boards/saml21-xpro/include/periph_conf.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+#define UART_0_PINS         (((PORT_PA22 | PORT_PA23) >> 16) | PORT_WRCONFIG_HWSEL)
+#define UART_0_REF_F        (4000000UL)
+
+/**
+ * @name SPI configuration
+ * @{
+ */
+#define SPI_NUMOF          (1)
+#define SPI_0_EN           1
+
```
Comment (haukepetersen) [inline boards/saml21-xpro/include/periph_conf.h]: `/** @} */` missing

File: boards/saml21-xpro/include/periph_conf.h [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * @name I2C configuration
+ * @{
+ */
+#define I2C_NUMOF          (0)
+
```
Comment (haukepetersen) [inline boards/saml21-xpro/include/periph_conf.h]: `/** @} */ missing

File: cpu/saml21/cpu.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    /* disable the watchdog timer */
+    WDT->CTRLA.bit.ENABLE = 0;
+
+    /* set pendSV interrupt to lowest possible priority */
+    NVIC_SetPriority(PendSV_IRQn, 0xff);
+
+    /* turn on MCLK */
+    MCLK->APBAMASK.reg |= MCLK_APBAMASK_GCLK;
+
+    /*OSCCTRL->OSC16MCTRL.reg |= OSCCTRL_OSC16MCTRL_FSEL(OSCCTRL_OSC16MCTRL_FSEL_16) |
```
Comment (haukepetersen) [inline cpu/saml21/cpu.c]: if you don't use this, throw it out...

File: cpu/saml21/cpu.c [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    GCLK->CTRLA.reg = GCLK_CTRLA_SWRST;
+    while (GCLK->CTRLA.reg & GCLK_CTRLA_SWRST);
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_SWRST);
+
+    /* Select the correct generator */
+    *((uint8_t*)&GCLK->GENCTRL[0].reg) = 0;
+    while (GCLK->SYNCBUSY.reg & GCLK_SYNCBUSY_SWRST);
+    GCLK->GENCTRL[0].reg = (
+            GCLK_GENCTRL_GENEN          /* enable gclk */
+            | GCLK_GENCTRL_SRC_OSC16M   /* select 16mhz osc */
```
Comment (haukepetersen) [inline cpu/saml21/cpu.c]: actually, I would strongly opt for initializing the CPU to its maximum possible frequency (48MHz?) for consistency reasons (we do this for every board in RIOT). In my opinion the paradigm should be: start out with the max clock and then reduce this based on your application needs if you really need to save power...

**Comment by haukepetersen** [comment]
ok, this PR is a bitch to review (too many large header files). Let's do the following: you fix the few comments I had and show me that the board is working and we merge this PR. Then we can fix any stuff that we find afterwards.

**Comment by kaspar030** [comment]
- accidently rebased
- comments addressed

Somehow I can't get the PLL to clock the board with 48MHz. But I've set the speed to 16MHz, can we keep that for now?

**Comment by haukepetersen** [comment]
Fine for now I guess. I have it on my list to look at the PLL stuff, maybe in the end of this week...

**Comment by kaspar030** [comment]
@haukepetersen *ping* You've seen this working, right?

**Comment by OlegHahm** [comment]
@haukepetersen, has probably still >800 GItHub notifications. No chance to trigger him this way. ;)

**Comment by haukepetersen** [comment]
Yes, I have seen this working. And there are still open comments above :-)

**Comment by kaspar030** [comment]
@haukepetersen The leds are fine now. Anything else?

**Comment by kaspar030** [comment]
Hm, follow-up PR's are coming. @OlegHahm, if I give you remote access, would you take a look?

**Comment by OlegHahm** [comment]
Sure - will be difficult to test the leds though. ;)

**Comment by haukepetersen** [comment]
All good from my side: ACK

**Comment by kaspar030** [comment]
- squashed
- rebased
Waiting for travis...

**Comment by kaspar030** [comment]
- fixed rebase introduced include bug

**Comment by haukepetersen** [comment]
please squash (again), ACK holds

**Comment by kaspar030** [comment]
- squashed

**Comment by haukepetersen** [comment]
Then lets wait for Travis and merge!

**Comment by kaspar030** [comment]
actually there's a problem with SPI at 16mhz, works fine with 4mhz. Its weird, because the formula for calculating the baudrate is so simple. I only tested initialization of ng_at86rf212b, maybe something is going too fast at 16mhz... Still I'd like to merge and find a solution an another PR.

**Comment by gebart** [comment]
The at86rf212b is picky with the timings between the bytes as well, not
just the baud rate. Check the data sheet for details on that.

**Comment by OlegHahm** [comment]
C++ header guards are missing in `cpu/saml21/include/atmel/component`

**Comment by kaspar030** [comment]
> C++ header guards are missing in cpu/saml21/include/atmel/component

I solved this by putting the #include of the first atmel header within ```#ifdef C++``` in cpu-conf.h.
That way we don't have to change all the headers. They are not supposed to be included by anything other than cpu-conf.h.

**Comment by OlegHahm** [comment]
Maybe you can modify the script, to exclude certain directories from the check? I just want to make sure, that the other Travis scripts are ran and show no errors.

**Comment by kaspar030** [comment]
I'll take a look...

**Comment by kaspar030** [comment]
- added the atmel dirs to externc's ignore regex

**Comment by OlegHahm** [comment]
Doesn't seem very scalable :wink:, but will work for now and is out of scope for thie PR.

**Comment by kaspar030** [comment]
@gebart Found it, it was spi_transfer_byte not waiting for receive to finish when ```in==null```, causing the CS line to be dropped too early. I wonder why this works in the very similar samr21 implementation...

Will commit in a follow up, I'm tired of rebasing & squashing & waiting *again* for travis. :)

**Comment by OlegHahm** [comment]
Somehow your last commit keeps on causing Travis to hang. I have no clue why.

**Comment by OlegHahm** [comment]
Runs without problems on my laptop though, apart from
```
cpu/saml21/syscalls.c:233: style (unreadVariable): Variable 'c' is assigned a value that is never used.
```

Merge at will.

**Comment by kaspar030** [comment]
> Somehow your last commit keeps on causing Travis to hang. I have no clue why.

seems like cppcheck takes too long on the atmel headers... I'll add an exclude mechanism to cppcheck.

**Comment by kaspar030** [comment]
- rebased
- exclude atmel headers also from cppcheck

**Comment by OlegHahm** [comment]
```
cpu/saml21/periph/gpio.c:345: performance (redundantAssignment): Variable 'res' is reassigned a value before the old one has been used.

cpu/saml21/periph/uart.c:90: style (unreadVariable): Variable 'baud_calculated' is assigned a value that is never used.

cpu/saml21/syscalls.c:233: style (unreadVariable): Variable 'c' is assigned a value that is never used.
```

**Comment by OlegHahm** [comment]
And most applications faill to build.

**Comment by kaspar030** [comment]
@OlegHahm: adaption to the new THREAD_defines was missing. Compile-test works flawlessly now.

**Comment by kaspar030** [comment]
We could save travis the work...

**Comment by kaspar030** [comment]
@OlegHahm The syscalls will be dropped in a follow-up PR.

**Comment by haukepetersen** [comment]
Compile tests work here as well. ACK and go.

**Pull request merged**